File: sdk/csharp/Ferrum.Sdk/Scale.cs
using System.Numerics;
using System.Text;

namespace Ferrum.Sdk;

/// <summary>
/// Minimal SCALE encoder for the Ferrum call types. Substrate.NetApi handles
/// transport and signing; these helpers produce the call-parameter bytes so the
/// SDK stays a thin, correct wrapper that cannot drift from a stale codegen.
/// </summary>
public sealed class ScaleWriter
{
    private readonly List<byte> _buf = new();

    public byte[] ToArray() => _buf.ToArray();

    public ScaleWriter U8(byte v) { _buf.Add(v); return this; }
    public ScaleWriter Bool(bool v) { _buf.Add((byte)(v ? 1 : 0)); return this; }

    public ScaleWriter U32(uint v) { _buf.AddRange(BitConverter.GetBytes(v)); return this; } // LE
    public ScaleWriter U64(ulong v) { _buf.AddRange(BitConverter.GetBytes(v)); return this; }

    public ScaleWriter U128(BigInteger v)
    {
        if (v.Sign < 0) throw new ArgumentException("u128 must be non-negative");
        var bytes = v.ToByteArray(isUnsigned: true, isBigEndian: false);
        if (bytes.Length > 16) throw new ArgumentException("value exceeds u128");
        _buf.AddRange(bytes);
        _buf.AddRange(new byte[16 - bytes.Length]); // pad to 16 LE
        return this;
    }

    /// <summary>Raw fixed-width bytes ([u8; N]) — no length prefix.</summary>
    public ScaleWriter Fixed(byte[] v) { _buf.AddRange(v); return this; }

    /// <summary>SCALE compact-encoded unsigned integer.</summary>
    public ScaleWriter Compact(BigInteger v)
    {
        if (v < 0) throw new ArgumentException("compact must be non-negative");
        if (v < 64) { _buf.Add((byte)(v << 2)); return this; }
        if (v < 16384) { uint x = (uint)((v << 2) | 1); _buf.AddRange(BitConverter.GetBytes((ushort)x)); return this; }
        if (v < 1073741824) { uint x = (uint)((v << 2) | 2); _buf.AddRange(BitConverter.GetBytes(x)); return this; }
        var bytes = v.ToByteArray(isUnsigned: true, isBigEndian: false);
        _buf.Add((byte)(((bytes.Length - 4) << 2) | 3));
        _buf.AddRange(bytes);
        return this;
    }

    /// <summary>Length-prefixed byte vector (Vec&lt;u8&gt; / BoundedVec&lt;u8&gt; / Bytes).</summary>
    public ScaleWriter Bytes(byte[] v) { Compact(v.Length); _buf.AddRange(v); return this; }

    /// <summary>ASCII string as a length-prefixed byte vector (BoundedVec&lt;u8&gt;).</summary>
    public ScaleWriter AsciiVec(string s) => Bytes(Encoding.ASCII.GetBytes(s));

    /// <summary>Fixed-width ASCII code ([u8;N]) — length-checked, no prefix.</summary>
    public ScaleWriter AsciiFixed(string s, int len)
    {
        var b = Encoding.ASCII.GetBytes(s);
        if (b.Length != len) throw new ArgumentException($"code '{s}' must be exactly {len} ASCII bytes");
        return Fixed(b);
    }

    /// <summary>Enum/variant discriminant byte followed by any inline fields.</summary>
    public ScaleWriter Variant(byte index) => U8(index);

    /// <summary>Vec&lt;T&gt; with a custom element writer.</summary>
    public ScaleWriter Vec<T>(IReadOnlyList<T> items, Action<ScaleWriter, T> writeElem)
    {
        Compact(items.Count);
        foreach (var it in items) writeElem(this, it);
        return this;
    }

    /// <summary>Option&lt;T&gt;: None=0x00, Some=0x01 then the value.</summary>
    public ScaleWriter Option<T>(T? value, Action<ScaleWriter, T> writeSome) where T : struct
    {
        if (value is null) return U8(0);
        U8(1);
        writeSome(this, value.Value);
        return this;
    }
}

public static class Hex
{
    public static byte[] Decode(string hex)
    {
        var s = hex.StartsWith("0x") ? hex[2..] : hex;
        if (s.Length % 2 != 0) throw new ArgumentException("odd-length hex");
        var bytes = new byte[s.Length / 2];
        for (int i = 0; i < bytes.Length; i++) bytes[i] = Convert.ToByte(s.Substring(i * 2, 2), 16);
        return bytes;
    }

    public static byte[] Bytes32(string hex)
    {
        var b = Decode(hex);
        if (b.Length != 32) throw new ArgumentException($"expected 32 bytes, got {b.Length}");
        return b;
    }

    /// <summary>Perbill parts-per-billion from a fraction of one (0..1).</summary>
    public static uint Perbill(double frac)
    {
        if (frac is < 0 or > 1) throw new ArgumentException("perbill fraction must be within [0,1]");
        return (uint)Math.Round(frac * 1_000_000_000.0);
    }
}

File: apps/federation-console/src/scale.js
// Minimal SCALE encoder — a byte-for-byte port of the C# SDK's ScaleWriter
// (sdk/csharp/Ferrum.Sdk/Scale.cs). Produces the same call-parameter bytes so
// extrinsics built here are identical to the reference SDK.
import { hexToU8a, stringToU8a, u8aConcat } from '@polkadot/util'

export class ScaleWriter {
  constructor() {
    this.parts = []
  }
  _push(u8a) {
    this.parts.push(u8a)
    return this
  }
  toU8a() {
    return u8aConcat(...this.parts)
  }

  u8(v) {
    return this._push(Uint8Array.of(v & 0xff))
  }
  bool(v) {
    return this.u8(v ? 1 : 0)
  }
  // little-endian fixed-width integers
  u32(v) {
    return this._push(leBytes(BigInt(v), 4))
  }
  u64(v) {
    return this._push(leBytes(BigInt(v), 8))
  }
  u128(v) {
    const b = BigInt(v)
    if (b < 0n) throw new Error('u128 must be non-negative')
    if (b >= 1n << 128n) throw new Error('value exceeds u128')
    return this._push(leBytes(b, 16))
  }
  // raw fixed-width bytes ([u8; N]) — no length prefix
  fixed(u8a) {
    return this._push(asU8a(u8a))
  }
  // SCALE compact-encoded unsigned integer
  compact(v) {
    let n = BigInt(v)
    if (n < 0n) throw new Error('compact must be non-negative')
    if (n < 64n) return this.u8(Number(n << 2n))
    if (n < 16384n) return this._push(leBytes((n << 2n) | 1n, 2))
    if (n < 1073741824n) return this._push(leBytes((n << 2n) | 2n, 4))
    const bytes = leTrim(n)
    this.u8(((bytes.length - 4) << 2) | 3)
    return this._push(bytes)
  }
  // length-prefixed byte vector (Vec<u8> / BoundedVec<u8> / Bytes)
  bytes(u8a) {
    const b = asU8a(u8a)
    this.compact(b.length)
    return this._push(b)
  }
  // ASCII string as a length-prefixed byte vector
  asciiVec(s) {
    return this.bytes(stringToU8a(s))
  }
  // fixed-width ASCII code ([u8; N]) — length-checked, no prefix
  asciiFixed(s, len) {
    const b = stringToU8a(s)
    if (b.length !== len) throw new Error(`code '${s}' must be exactly ${len} ASCII bytes`)
    return this.fixed(b)
  }
  // Vec<T> with a custom element writer
  vec(items, writeElem) {
    this.compact(items.length)
    for (const it of items) writeElem(this, it)
    return this
  }
}

// ---- helpers -------------------------------------------------------------
function leBytes(value, len) {
  const out = new Uint8Array(len)
  let v = BigInt(value)
  for (let i = 0; i < len; i++) {
    out[i] = Number(v & 0xffn)
    v >>= 8n
  }
  return out
}

function leTrim(value) {
  let v = BigInt(value)
  const out = []
  if (v === 0n) return Uint8Array.of(0)
  while (v > 0n) {
    out.push(Number(v & 0xffn))
    v >>= 8n
  }
  return Uint8Array.from(out)
}

function asU8a(x) {
  if (x instanceof Uint8Array) return x
  if (typeof x === 'string') return hexToU8a(x.startsWith('0x') ? x : '0x' + x)
  if (Array.isArray(x)) return Uint8Array.from(x)
  throw new Error('expected bytes')
}

// 32-byte hash from hex, length-checked.
export function bytes32(hex) {
  const b = asU8a(hex)
  if (b.length !== 32) throw new Error(`expected 32 bytes, got ${b.length}`)
  return b
}

// Perbill (parts-per-billion) from a percentage (0..100).
export function perbillFromPercent(pct) {
  const p = Number(pct)
  if (p < 0 || p > 100) throw new Error('percent must be within [0,100]')
  return Math.round((p / 100) * 1_000_000_000)
}

export { asU8a }
